### ISSUE TITLE
Bugzilla Wrapper using Perceval

### ISSUE THREAD

**Issue by leilani-reich** [sailuh/kaiaulu#152]
We would like to implement a wrapper to parse Bugzilla data. This will be implemented similarly to how Kaiaulu parses gitlogs, and we will use Perceval to parse the data.

The main contributions for this issue will be the Bugzilla wrapper function itself, which will be added as a function in parser.R; unit tests for this function; and a vignette demonstrating how to use the Bugzilla wrapper.

Deadline Tuesday

Leilani:
Pull Request: https://github.com/sailuh/kaiaulu/pull/155
- [x] Bugzilla download function
- [x] Bugzilla parse function using traditional Perceval Bugzilla backend
- [x] Bugzilla parse function using REST API Perceval Bugzilla backend

Malia:
Pull Request: https://github.com/sailuh/kaiaulu/pull/164
- [x] Bugzilla vignette showing use of Bugzilla functions

Nico:

- [ ] Unit Testing for Bugzilla functions

**Comment by carlosparadis** [comment]
Thanks! Please note Perceval has two endpoints for Bugzilla. Could you take a look at why that is the case? I.e. can you identify one project that requires each endpoint? This will help inform #153 task:

https://github.com/chaoss/grimoirelab-perceval#bugzilla

I'd also encourage you to try and expand a bit more on this issue about the output you obtain in json, one example output, and how you plan to name the columns after converting it to a table in the parser function.

**Comment by leilani-reich** [comment]
Hi Carlos, I saw the two different endpoints for Bugzilla, one using the traditional backend, and the other using the REST API backend. This is the case because the REST API function can work much faster to retrieve data for Bugzilla servers that are version 5.0 or higher. One project that looks like it requires the traditional backend is the Netfilter Project (https://bugzilla.netfilter.org/) since it has version 4.4.13 of Bugzilla. Meanwhile, one project that can use the REST API is the Apache Project (https://bz.apache.org/bugzilla/). 

The output I obtain in JSON from calling the traditional Perceval function to get data from Bugzilla is quite messy. I want to extract the key fields containing the information relevant to a bug report and use those fields as columns in my data.table. I want to rename the columns into a simpler form that the user can understand.

In particular, I have used the columns "bug_id", "version", "product", "component", "assigned_user", "assigned_name", "severity","status", "summary", "date_created", "date_modified" in the tables I have created thus far. 

The original names for these columns were "data.bug_id", "data.version", "data.product", "data.component", "assigned_to" (which I split into assigned_user & assigned_name), "data.bug_severity", "data.bug_status", "data.short_desc", "data.creation_ts", and "data.delta_ts."

Here is an example output:

| bug_id<chr> | version<chr> | product<chr> | component<chr> | assigned_user<chr> |  assigned_name<chr> | severity<chr> | status | summary | date_created<chr> | date_modified<chr>
| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
| 66457 | 1.7.2 | APR | APR test | bugs | Apache Portable Runtime bugs mailinglist | normal | NEW | testatomic: Line 908: Unexpected value | 2023-02-02 01:55:33 +0000 | 2023-02-10 01:32:28 +0000

Also, here is my rough code for the function itself:

```{r}
#' Parse bugzilla from Perceval
#'
#' @param perceval_path path to perceval binary
#' @param bugzilla_link link to specific bugzilla site
#' @param save_path optional save path for .rds object
#' @export
#' @family parsers
parse_bugzilla <- function(perceval_path,bugzilla_link,date,save_path=NA){
 
  # Expand paths (e.g. "~/Desktop" => "/Users/someuser/Desktop")
  perceval_path <- path.expand(perceval_path)
  save_path <- ifelse(!is.na(save_path),path.expand(save_path),NA)

  # Parsed JSON output. 
  # Matches command in perceval to get bugzilla data via the traditional backend
  perceval_output <- system2(perceval_path,
                             args = c('bugzilla',bugzilla_link,'--json-line','--from-date',date),
                             stdout = TRUE,
                             stderr = FALSE)

  # Create the data.table from JSON output
  perceval_parsed <- data.table(jsonlite::stream_in(textConnection(perceval_output),verbose = FALSE))

  # To learn about the data
  # print(names(perceval_parsed))
  
  # Here, I take the most relevant columns of information (other columns get dropped)
  # There are a lot of 1x1 dataframes that must be unlisted first, which is done here
  perceval_parsed <- perceval_parsed[, .(bug_id=unlist(data.bug_id), 
                                          version=unlist(data.version),
                                          component=unlist(data.component),
                                          product=unlist(data.product),
                                          severity=unlist(data.bug_severity),
                                          status=unlist(data.bug_status),
                                          summary=unlist(data.short_desc),
                                          date_created=unlist(data.creation_ts),
                                          date_modified=unlist(data.delta_ts),
                                          data.assigned_to = data.assigned_to
                                         )
                                        ]
  
  # Unlisting other columns which are not 1x1 dataframes and need more work
  perceval_parsed <- perceval_parsed[, .(
                                          assigned_user=unlist(data.assigned_to[[1]]["__text__"]),
                                          # Set name field to NA if the name isn't found
                                      assigned_name=append(unlist(data.assigned_to[[1]]),NA)[2]),,
                                                                                              by=list(
                                                                                                bug_id,
                                                                                                version,
                                                                                              product,
                                                                                              component,
                                                                                              severity,
                                                                                              status,
                                                                                              summary,
                                                                                              date_created,
                                                                                              date_modified)]
                                     
  
  # Reorder the columns in a sensible format
  perceval_parsed <- data.table::setcolorder(perceval_parsed, c("bug_id", "version", "product", "component", "assigned_user", "assigned_name", "severity","status", "summary", "date_created", "date_modified"))

  #Parsing bugzilla can take a while, save if a path is provided
   if(!is.na(save_path)){
     saveRDS(perceval_parsed,save_path)
   }
  return(perceval_parsed)
}

bugzilla_link <- "https://bugzilla.kernel.org"
perceval_path <- "path/to/perceval"
date <- "2023-02-10"
save_path <- "path/to/saveRDS"

parse_bugzilla(perceval_path, bugzilla_link, date, save_path)
```

I noticed this function runs very slow depending on the bugzilla site and the date you set.

Please let me know how I can improve. Thank you.

**Comment by carlosparadis** [comment]
Hi,

1. Just to make sure, could you confirm the Perceval version you used for this? We want to be consistent with the older version. 

2. Also, on GitHub, to highlight code in a block use  r instead of{r} (which is the RStudio notation). I made a minor edit for your message so you can see what it looks like if you enter edit mode on it. 

2. Regarding what information is being tabulated, I believe it would be helpful if, on Tuesday's meeting, you could present it to @rnkazman and also mention here what data is in there, and which data you chose to tabulate so we can also consider if any other is needed. Could you compare to the Jira notebook to see if we are missing any columns or have any additional ones?

 
3. As far as column names are concerned, could you check the parse_jira function and ensure the columns that convey the same data match the nomenclature on it? 

http://itm0.shidler.hawaii.edu/kaiaulu/reference/parse_jira.html 

Please also see the table here: http://itm0.shidler.hawaii.edu/kaiaulu/articles/download_jira_data.html

4. For JIRA, there was one function to download the raw data as json, and another to parse said json. I see with Perceval you have a parse_and_download function, rather than just parse. Is there any way to de-couple that in perceval? I.e. call it to first download the raw data, then call it to parse said data? Maybe this would speed the processing. 

5. I am not sure which of the two endpoints in Perceval you wrapped around. The one before, or the one after REST API?

6. Is there any means to specify the time window one wishes to download? 

7. What is the naming convention you are using when saving the file? 

8. Any other assumptions made? 

9. How slow is slow? One hour to download one year of issues? 5 hours? A whole day?

Thanks!

**Comment by leilani-reich** [comment]
1. I used Perceval 0.12.24.
2. Thanks for telling me. Wasn't sure what to put for this.
3. Understood. I did notice some columns missing between my table and the JIRA table, as well as extra columns my table had that the JIRA table did not. I have made some changes.

The common columns I have gotten between the Bugzilla columns and JIRA columns are as follows:
"issue_key", "issue_version","issue_product","issue_component","issue_assignee_id","issue_assignee_name",
             "issue_severity", "issue_status", "issue_summary","issue_description","issue_created_datetimetz", 
             "issue_updated_datetimetz", "issue_reporter_id", "issue_reporter_name", "issue_type"

I've also identified extra columns which may be useful for Bugzilla specifically, based on what I've seen from looking at the advanced search query filters on different Bugzilla sites, such as [here](https://bz.apache.org/bugzilla/query.cgi?format=advanced).

I will make sure my naming conventions for columns aligns with the JIRA columns.

4. Yes, I was able to download the raw data first and parse it separately.

This is how I downloaded the data:
```r
perceval_path <- path/to/perceval
bugzilla_link <- https://bugzilla-site.com
date <- '2023-02-19'

# To download the data
json_issues <- system2(perceval_path,
                             args = c('bugzilla',bugzilla_link,'--json-line','--from-date',date),
                             stdout = TRUE,
                             stderr = FALSE)
```

This did seem to speed the processing.

5. The one before the REST API? I used the traditional backend "bugzilla" shown [here](https://github.com/chaoss/grimoirelab-perceval#bugzilla) to help implement the parser.
6. The user can specify the date to start the bug retrieval, but I don't think they can specify an exact time window.
7. When saving the file, I called it project_bugzilla.RDS. This matches the project_mbox, project_jira, etc in the Social Smells notebook.
8. I assume the user does not care about logging in. There is an option to provide a username and password for the Perceval Bugzilla function, but it seems unnecessary.
9. For https://bugzilla.kernel.org/, it took around an hour and a half to download a years worth of issues, which was about 2500 bugs. However, the speed of the function depends on how many bugs a site contains. For instance, https://bugzilla.redhat.com/ has a lot more bug reports daily, which makes bug retrieval much slower for the same time frame.


Also, do you want a download comments section for the Bugzilla function like there is for JIRA [here](http://itm0.shidler.hawaii.edu/kaiaulu/articles/download_jira_data.html#download-issues-without-comments-)?

Thank you.

**Comment by leilani-reich** [comment]
Hi Carlos,
I didn't see a way to just download Bugzilla comments only using Perceval, so I didn't make a download_bugzilla_comments() function.

However, I've created a new function called parse_bugzilla_issues_comments() to parse Bugzilla issues and comments. It is similar to the parse_jira() function in Kaiaulu.

## Here are the differences between the columns used in the Jira tables and my Bugzilla tables:

### Jira Issue Table:
| issue_key | issue_summary	| issue_type | issue_status | issue_resolution | issue_description | issue_created_datetimetz | issue_updated_datetimetz | issue_resolution_datetimetz | issue_creator_id | issue_creator_name | issue_creator_timezone |	issue_reporter_id | issue_reporter_name | issue_reporter_timezone | issue_components | issue_assignee_id | issue_assignee_name | issue_assignee_timezone |
| - | - | - | - | - | - | - | - | - | - | - | - | - | - | - | - | - | - | - |

### Bugzilla Issue Table:
| issue_key | issue_summary	| issue_type | issue_status | issue_resolution | issue_description | issue_created_datetimetz | issue_updated_datetimetz | issue_reporter_id | issue_reporter_name | issue_components | issue_assignee_id | issue_assignee_name | issue_target_milestone | issue_rep_platform | issue_status_whiteboard |
| - | - | - | - | - | - | - | - | - | - | - | - | - | - | - | - |

## Here is the function I made to download the Bugzilla data into a json file:

```r
#'  Download Bugzilla issues and comments using Perceval
#'
#' @param perceval_path path to perceval binary
#' @param bugzilla_site link to specific bugzilla site
#' @param date the date to start bug retrieval
#' @export
download_bugzilla <- function(perceval_path, bugzilla_site, date){
  json_issues <- system2(perceval_path,
                             args = c('bugzilla',bugzilla_site,'--json-line','--from-date',date),
                             stdout = TRUE,
                             stderr = FALSE)
  return(json_issues)
}

# Function call example
json_issues <- download_bugzilla("path/to/perceval", "https://bugzilla.site.com", "2023-02-22")

```



## Here is the parse_bugzilla_issues_comments() function:

```{r}
#' Parse Bugzilla issues and comments using Perceval
#'
#' @param bugzilla_json json object containing bugzilla issue data
#' @export
#' @family parsers
parse_bugzilla_issues_comments <- function(bugzilla_json){
  
  json_issue_comments <- data.table(jsonlite::stream_in(textConnection(bugzilla_json),verbose = FALSE))
  
  # To learn about the data
  #print(names(json_issue_comments))
  
  # # Comments list parser. Comments may occur on any json issue.
  bugzilla_parse_comment <- function(comment){
    # Get the number of comments for an issue. This count includes the issue desc. 
    num_comments <- length(comment[["commentid"]])
    
    # First comment is issue description, so we start indexing at 2
    parsed_comment <- list()
    parsed_comment[["comment_id"]] <- append(comment[["commentid"]][[2]],NA)[[1]]
    parsed_comment[["comment_author_id"]] <- append(comment[["who"]][[2]],NA)[[1]]
    parsed_comment[["comment_author_name"]] <- append(comment[["who"]][[2]],NA)[[2]]  
    parsed_comment[["comment_body"]] <- append(comment[["thetext"]][[2]],NA)[[1]]  
    parsed_comment[["comment_created_datetimetz"]] <- append(comment[["bug_when"]][[2]],NA)[[1]]
    parsed_comment[["comment_count"]] <- append(comment[["comment_count"]][[2]],NA)[[1]]
    parsed_comment[["comment_is_private"]] <- append(comment[["isprivate"]][[2]],NA)[[1]]
    
    parsed_comments <- list()
    # I need to make parsed_comments a list when I first add to it or else the table won't be correct
    parsed_comments <- append(list(parsed_comments), list(parsed_comment))
    
    # If there is more than one comment for an issue, parse them
    if (num_comments > 2) {
      
      for (i in 3:num_comments){
        parsed_comment <- list()
        
        parsed_comment[["comment_id"]] <- append(comment[["commentid"]][[i]],NA)[[1]]
        
        parsed_comment[["comment_author_id"]] <- append(comment[["who"]][[i]],NA)[[1]]
        parsed_comment[["comment_author_name"]] <- append(comment[["who"]][[i]],NA)[[2]]
    
        parsed_comment[["comment_body"]] <- append(comment[["thetext"]][[i]],NA)[[1]]
        
        parsed_comment[["comment_created_datetimetz"]] <- append(comment[["bug_when"]][[i]],NA)[[1]]
        parsed_comment[["comment_count"]] <- append(comment[["comment_count"]][[i]],NA)[[1]]
        parsed_comment[["comment_is_private"]] <- append(comment[["isprivate"]][[i]],NA)[[1]]
        
        parsed_comments <- append(parsed_comments, list(parsed_comment))
      }
    }
    
    return(parsed_comments)
  }
  
  # Number of issues in table
  n_issues <- length(unique(json_issue_comments[["data.bug_id"]]))
  #print(n_issues)
  
  # Prepare two lists which will contain data.tables for all issues and all comments
  # Both tables can share the issue_key, so they can be joined if desired.
  all_issues <- list()
  all_issues_comments <- list()
  
  for(i in 1:n_issues){
    
    # Obtain the issue_key
    issue_key <- json_issue_comments[["data.bug_id"]][[i]][[1]]
    
    issue_comment <- json_issue_comments
    
    # Parse all relevant *issue* fields
    all_issues[[i]] <- data.table(
      issue_key = issue_key,
      
      issue_summary = append(issue_comment[["data.short_desc"]][[i]],NA)[1],
      issue_type = append(issue_comment[["category"]][[i]],NA)[1],
      issue_status = append(issue_comment[["data.bug_status"]][[i]],NA)[1],
      issue_resolution = append(issue_comment[["data.resolution"]][[i]],NA)[1],
      issue_components = append(issue_comment[["data.component"]][[i]],NA)[1],
      issue_description = append(issue_comment[["data.long_desc"]][[i]][["thetext"]][[1]],NA)[1],
      
      issue_created_datetimetz = append(issue_comment[["data.creation_ts"]][[i]],NA)[1],
      issue_updated_datetimetz = append(issue_comment[["data.delta_ts"]][[i]],NA)[1],
      
      issue_assignee_id = append(issue_comment[["data.assigned_to"]][[i]][["__text__"]],NA)[1],
      issue_assignee_name = append(issue_comment[["data.assigned_to"]][[i]][["name"]],NA)[1],
      
      issue_reporter_id = append(issue_comment[["data.reporter"]][[i]][["__text__"]],NA)[1],
      issue_reporter_name = append(issue_comment[["data.reporter"]][[i]][["name"]],NA)[1],
      
      issue_target_milestone = append(issue_comment[["data.target_milestone"]][[i]],NA)[1],
      issue_rep_platform = append(issue_comment[["data.rep_platform"]][[i]],NA)[1],
      issue_status_whiteboard = append(issue_comment[["data.status_whiteboard"]][[i]],NA)[1]
    )
    
    # Comments
    # For each issue, it may have 0 or more comments (not including issue desc). Parse them
    # in a separate table.
    
    root_of_comments_list <- json_issue_comments[["data.long_desc"]][[i]]
    #print(root_of_comments_list)
    #If root_of_comments_list does not exist, then this is an issue only json, skip parsing
    if(length(root_of_comments_list) > 0){
      comments_list <- json_issue_comments[["data.long_desc"]][[i]]
      
      # Even on a json with comments, some issues may not have comments, check if comments exist:
      if(length(comments_list[["commentid"]]) > 1){
        # Parse all comments into issue_comments
        issue_comments <- rbindlist(bugzilla_parse_comment(comments_list))
        # Add issue_key column to the start of the table
        issue_comments <- cbind(data.table(issue_key=issue_key),issue_comments)
        all_issues_comments[[i]] <- issue_comments
      }
    }
  }
  
  all_issues <- rbindlist(all_issues,fill=TRUE)
  all_issues_comments <- rbindlist(all_issues_comments,fill=TRUE)
  
  parsed_issues_comments <- list()
  parsed_issues_comments[["issues"]] <- all_issues
  parsed_issues_comments[["comments"]] <- all_issues_comments
  
  return(parsed_issues_comments)
}

# Function call example
bugzilla_comments_table <- parse_bugzilla_comments(json_issues)

# Retrieve issues and comments separately
bugzilla_issues <- bugzilla_comments_table[["issues"]]
bugzilla_comments <- bugzilla_comments_table[["comments"]]
```

Thanks,
Leilani

**Comment by carlosparadis** [comment]
Since we followed on call this: https://github.com/sailuh/kaiaulu/issues/152#issuecomment-1436782015, I will skip my reply to your next comment. 

Could you make the tables vertical? You want two tables, of two columns each. E.g.

|Jira Column Name|BugZilla Column Name| 
|------------------|-----------------------|
|issue_key|issue_key| 

One for the issue table, and one for the comment table. The current format is hard to compare side by side. 

-------

I've made a minor edit for your second code block since you used {r} on it, just to fix the highlight. I think at this point it is a bit hard to comment the code at issue level, so I'd recommend you just open a pull request tied to this issue, and I can give you inline code feedback as I've recommended it could be done for Nicholas grammar fix suggestions.  

The function including issues and comments is fine. Please let me know if you could implement both Perceval endpoints. 

Will Nicholas be the one adding unit tests for this? 

So the functions appear on the documentation in itm0 under References, please also take a look on kaiaulu folder for the file _pkgdown.yml, and how new function names are added. It just be just a few lines (compare to how the existing function names are listed in References in itm0).  In retrospect, I may have missed that in CONTRIBUTING.md. Moreover, the NEWS.md would be edited in the major feature section. 

Lastly, since this is a code contribution, please go ahead and include your name in the description file too. 

Thank you!

**Comment by leilani-reich** [comment]
Hi Carlos, I just made a pull request.

In regards to the REST API Bugzilla endpoint, I just made another function for that. So there are two functions for parsing: parse_bugzilla (uses traditional backend) and parse_bugzillarest (uses REST API backend)

@nicoelee123  will be working on the unit test for this issue, yes.

I made the appropriate changes to _pkgdown.yml, NEWS.md, and DESCRIPTION.

Also, I've updated the tables in my last comment into a vertical column format, which will be easier to compare.

Thanks
